pkg/ingester/stream.go

```diff
@@ -173,132 +177,171 @@ func (s *stream) Push(
+	var (
+		outOfOrderSamples, outOfOrderBytes   int
+		rateLimitedSamples, rateLimitedBytes int
+		totalBytes                           int
+		failedEntriesWithError               []entryWithError
+		limit                                = s.limiter.lim.Limit()
+		lastLine                             = s.lastLine
+		highestTs                            = s.highestTs
+		toStore                              = make([]logproto.Entry, 0, len(entries))
+	)
+
+	for i := range entries {
+		// If this entry matches our last appended line's timestamp and contents,
+		// ignore it.
+		//
+		// This check is done at the stream level so it persists across cut and
+		// flushed chunks.
+		//
+		// NOTE: it's still possible for duplicates to be appended if a stream is
+		// deleted from inactivity.
+		if entries[i].Timestamp.Equal(lastLine.ts) && entries[i].Line == lastLine.content {
+			continue
 		}
 
-		s.tailerMtx.RLock()
-		hasTailers := len(s.tailers) != 0
-		s.tailerMtx.RUnlock()
-		if hasTailers {
-			go func() {
-				stream := logproto.Stream{Labels: s.labelsString, Entries: storedEntries}
-
-				closedTailers := []uint32{}
-
-				s.tailerMtx.RLock()
-				for _, tailer := range s.tailers {
-					if tailer.isClosed() {
-						closedTailers = append(closedTailers, tailer.getID())
-						continue
-					}
-					tailer.send(stream, s.labels)
-				}
-				s.tailerMtx.RUnlock()
+		// The validity window for unordered writes is the highest timestamp present minus 1/2 * max-chunk-age.
+		cutoff := s.highestTs.Add(-s.cfg.MaxChunkAge / 2)
+		if !isReplay && s.unorderedWrites && !s.highestTs.IsZero() && cutoff.After(entries[i].Timestamp) {
+			failedEntriesWithError = append(failedEntriesWithError, entryWithError{entries[i], chunkenc.ErrTooFarBehind(cutoff)})
+			outOfOrderSamples++
+			outOfOrderBytes += len(entries[i].Line)
+			continue
+		}
 
-				if len(closedTailers) != 0 {
-					s.tailerMtx.Lock()
-					defer s.tailerMtx.Unlock()
+		totalBytes += len(entries[i].Line)
 
-					for _, closedTailerID := range closedTailers {
-						delete(s.tailers, closedTailerID)
-					}
-				}
-			}()
+		lastLine.ts = entries[i].Timestamp
+		lastLine.content = entries[i].Line
+		if highestTs.Before(entries[i].Timestamp) {
+			highestTs = entries[i].Timestamp
 		}
+
+		toStore = append(toStore, entries[i])
 	}
 
-	if len(s.chunks) != prevNumChunks {
-		s.metrics.memoryChunks.Add(float64(len(s.chunks) - prevNumChunks))
+	// Each successful call to 'AllowN' advances the limiter. With all-or-nothing
+	// ingestion, the limiter should only be advanced when the whole stream can be
+	// sent
+	now := time.Now()
+	if !s.limiter.AllowN(now, totalBytes) {
+		rateLimitedSamples = len(toStore)
+		rateLimitedBytes = totalBytes
+		for i := 0; i < len(toStore); i++ {
+			failedEntriesWithError = append(failedEntriesWithError, entryWithError{toStore[i], &validation.ErrStreamRateLimit{RateLimit: flagext.ByteSize(limit), Labels: s.labelsString, Bytes: flagext.ByteSize(len(toStore[i].Line))}})
```

Comment:
Ahhh, this is the reason `entryWithError` used a pointer. That way this slice only ever stores pointers to memory that already exists even though it may have to alloc to pass things around.